packages/belos/tpetra/test/MINRES/test_minres_diag.cpp

```diff
@@ -0,0 +1,418 @@
+      std::string opString="Iterative Solver", bool print_in=false);
+
+  virtual ~Iterative_Inverse_Operator() {}
+
+  virtual void operator () (const MV &b, MV &x);
+
+  private:
+
+  Teuchos::RCP<Vector_Operator> pA;       // operator which will be inverted
+  // supplies a matrix std::vector multiply
+  const bool print;
+
+  Teuchos::Time timer;
+  Teuchos::RCP<Teuchos::Comm<int>> pComm;
+  Teuchos::RCP<MP>  pMap;
+
+  Teuchos::RCP<OP> pPE;
+  Teuchos::RCP<Teuchos::ParameterList>         pList;
+  Teuchos::RCP<Belos::LinearProblem<double,MV,OP> >   pProb;
+  Teuchos::RCP<Belos::MinresSolMgr<double,MV,OP> >      pBelos;
+};
+
+Iterative_Inverse_Operator::Iterative_Inverse_Operator(int n_in, int blocksize,
+    const Teuchos::RCP<Vector_Operator>& pA_in,
+    std::string opString, bool print_in)
+: Vector_Operator(n_in, n_in),      // square operator
+  pA(pA_in),
+  print(print_in),
+  timer(opString)
+{
+
+  int n_global;
+// TD: could the following ifdef block be replaced by just pComm = Tpetra::getDefaultComm() ?
+// AM: yes
+#ifdef HAVE_TPETRA_MPI
+  MPI_Allreduce(&n, &n_global, 1, MPI_UNSIGNED, MPI_SUM, MPI_COMM_WORLD);
+  pComm = Teuchos::rcp (new Teuchos::MpiComm<int> (MPI_COMM_WORLD));
+#else
+  pComm = Teuchos::rcp (new Teuchos::SerialComm<int> ());
+  n_global = n;
+#endif
+  pMap =  Teuchos::rcp( new MP(n_global, n, 0, pComm) );
+
+  pPE = Teuchos::rcp( new Trilinos_Interface(pA, pComm, pMap ) );
+
+  pProb = Teuchos::rcp( new Belos::LinearProblem<double,MV,OP>() );
+  pProb->setOperator( pPE );
+
+  int max_iter = 100;
+  double tol = 1.0e-10;
+  int verbosity = Belos::Errors + Belos::Warnings;
+  if (print)
+    verbosity += Belos::TimingDetails + Belos::StatusTestDetails;
+
+  pList = Teuchos::rcp( new Teuchos::ParameterList );
+  pList->set( "Maximum Iterations", max_iter );
+  pList->set( "Convergence Tolerance", tol );
+  pList->set( "Verbosity", verbosity );
+
+  pBelos = Teuchos::rcp( new Belos::MinresSolMgr<double,MV,OP>(pProb, pList) );
+}
+
+void Iterative_Inverse_Operator::operator () (const MV &b, MV &x)
+{
+  int pid = pComm->getRank();
+
+  // Initialize the solution to zero
+  x.putScalar( 0.0 );
+
+  // Reset the solver, problem, and status test for next solve (HKT)
+  pProb->setProblem( Teuchos::rcp(&x, false), Teuchos::rcp(&b, false) );
+
+  timer.start();
+  Belos::ReturnType ret = pBelos->solve();
+  timer.stop();
+
+  if (pid == 0 && print) {
+    if (ret == Belos::Converged)
+    {
+      std::cout << std::endl << "pid[" << pid << "] Minres converged" << std::endl;
+      std::cout << "Solution time: " << timer.totalElapsedTime() << std::endl;
+
+    }
+    else
+      std::cout << std::endl << "pid[" << pid << "] Minres did not converge" << std::endl;
+  }
+}
+
+//************************************************************************************************
+//************************************************************************************************
+
+int main(int argc, char *argv[])
```

Comment:
How do we test for the different types?

Proposed solution to manage several types of tests:
```cpp
template <class Type>
void run(){
  using ST = Tpetra::MultiVector<Type, ...>::scalar_type;
  // run tests
}

int main(int argc, char *argv[]){
  run<double>();
  run<float>();
  // ...
}
```

Reply:
Good, implement it